### PR TITLE
Instrumentation Adapters to support Activity API

### PR DESCRIPTION
**Pull request by cijothomas**
Goal: Fixes 3-a,c,d,e,f from #684

An alternate proposal for https://github.com/open-telemetry/opentelemetry-dotnet/pull/692
There are quite a number of TODOs part of this PR. I want to address them outside of this PR, as this PR already got too much with all the instrumentationadapters!. Want to fix all instrumentionadapters so as to find any issues with Activity API and give .NET Team feedback on the same.

This approach does the following:
Adds `builder.AddActivitySource(string.Empty)` to get callbacks for Activities created by old instrumentations. (they do new Activity(), instead of using ActivitySource). The ActivitySource will be empty for the activities created using old approach, and hence `string.Empty` is subscribed to get them.

Uses the same activity created by existing instrumentation, and adds tags to it by reading from payload.
The instrumentation now makes calls to Sampler, and set `IsAllDataRequested` and TraceFlags on Activity appropriately.

Compared to #692 , this avoids creating a new duplicate Activity. But this now forces Instrumentation to have dependency on the SDK, to do sampling.

Major TODOs I'll address following this PR:
1. Kind will be settable once .NET ships new build. This PR currently uses reflection.

2. Avoid instrumentation being dependent on Sampling - its currently hardcoded. We are exploring if .NET can expose a method to run `GetRequestedDataUsingContext` pipeline (this is where we attach sampling logic) for activities created outside ActivitySource. If this happens, we'll simply switch to it. If not, we'll expose similar in OP.API, so instrumentation simply calls this method which internally can determine sampling. (Exact details to be designed as followup)

3. Subscribing to ActivitySource String.Empty is probably too noisy. Need to design an alternative. Its possible to NOT subscribe to old Activity using ActivitySource, and instead make Instrumentation call the Start/Stop (or equivalent). (Exact details to be designed as followup)

4. Support ITextFormat for ActivityContext. This PR simply ignores this code path now.

5. Find a way to do filters. The current approach of httpclient/aspnet/core Filter mechanism is broken with this PR.

6. AzureClient instrumentation - kept it currently, but this will be moved outside of this repo.

7. Move the diagnosticsource listener code from SDK to API, so Instrumentation can depend on it by just referring to the OT API. (This issue is already existing, but listing here as well)

8. Cleanup ListenerHandle - there is code copying activity tags into span etc - they need to be removed once Activity replaces Span.

(tests doing some of the above are disabled until actual code fix is done)

I will open issues to track these separately and submit follow ups.

### CODE REVIEW

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs [hunk 1]
```diff
@@ -60,10 +62,21 @@ public override void OnStartActivity(Activity activity, object payload)
...
 
+            // TODO: the line below once .NET ships new Activity
+            // Or do reflection now.
+            // activity.ActivityKind = ActivityKind.Server
+
             var request = context.Request;
 
             // see the spec https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-semantic-conventions.md
             var path = (request.PathBase.HasValue || request.Path.HasValue) ? (request.PathBase + request.Path).ToString() : "/";
+            if (activity != null)
```
Comment (tarekgh) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:73]: >activity [](start = 16, length = 8)

is it possible the Activity can be null?

Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:73]: No. I didnt need the activity null check.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs [hunk 2]
```diff
@@ -35,6 +36,7 @@ internal class HttpInListener : ListenerHandler
         private readonly PropertyFetcher beforeActionTemplateFetcher = new PropertyFetcher("Template");
         private readonly bool hostingSupportsW3C = false;
         private readonly AspNetCoreInstrumentationOptions options;
+        private readonly ActivitySampler sampler = new AlwaysOnActivitySampler();
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:39]: hardcoded sampler for just prototyping.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs [hunk 3]
```diff
@@ -60,10 +62,21 @@ public override void OnStartActivity(Activity activity, object payload)
...
             var request = context.Request;
 
             // see the spec https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-semantic-conventions.md
             var path = (request.PathBase.HasValue || request.Path.HasValue) ? (request.PathBase + request.Path).ToString() : "/";
+            if (activity != null)
+            {
+                activity.DisplayName = path;
+            }
+
+            var samplingDecision = this.sampler.ShouldSample(activity.Context, activity.TraceId, default, path, activity.Kind, activity.Tags, activity.Links);
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:78]: This is the part which makes me not 100% convinved this approach is right. All instrumentationadatpers now need to know about Samplers (which is SDK concept).

Instead of simply checking if activity.IsAllDataRequested, its now first setting IsAllDataRequested by running sampling algorithm

File: src/OpenTelemetry/Trace/Export/SimpleActivityProcessor.cs [hunk 4]
```diff
@@ -41,6 +43,16 @@ public SimpleActivityProcessor(ActivityExporter exporter)
         /// <inheritdoc />
         public override void OnStart(Activity activity)
         {
+            // Perform sampling decision if source is empty.
```
Comment (cijothomas) [inline src/OpenTelemetry/Trace/Export/SimpleActivityProcessor.cs:46]: this is not 100% correct, as the sampling decision is now made with the current ActivityKind (which would be Internal, the default value). Though none of the built-in Samplers makes use of ActivityKind, nothing prevents one from writing Sampler which takes Kind into consideration.

More better would be to "feed" this activity created outside of `ActivitySource`, into the `ActivityListener`'s `GetRequestedDataUsingContext` pipeline. That can be done in the instrumentationadapter itself, *after* setting the Kind to correct one, (and possibly enrich activity with more tags/links/events etc. if needed)

Comment (cijothomas) [reply src/OpenTelemetry/Trace/Export/SimpleActivityProcessor.cs:46]: Thanks @CodeBlanch for proposing the idea of "feeding" an external activity to the `GetRequestedDataUsingContext` pipeline.

File: src/OpenTelemetry/Trace/Export/SimpleActivityProcessor.cs [hunk 5]
```diff
@@ -41,6 +43,16 @@ public SimpleActivityProcessor(ActivityExporter exporter)
...
+            // This occurs for activities created outside of ActivitySource
+            // i.e new Activity() directly.
+            // Note: This also can occur if one creates an ActivitySource of the name string.empty.
+            // In that case, the sampling logic would be run twice!
+            if (activity.Source.Name.Equals(string.Empty))
```
Comment (cijothomas) [inline src/OpenTelemetry/Trace/Export/SimpleActivityProcessor.cs:51]: this also looks hacky, as the sampling algorithm would be run more than once, if one creates ActivitySource with the name `string.Empty`.

File: src/OpenTelemetry/Trace/Export/SimpleActivityProcessor.cs (hunk 4 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Trace/Export/SimpleActivityProcessor.cs:46]: Did you present the idea to .NET team? Any feedback?

Comment (cijothomas) [reply src/OpenTelemetry/Trace/Export/SimpleActivityProcessor.cs:46]: Given that making sampling decision in SDK piece is going to be incorrect (as we have incorrect Kind), I'll move the sampling logic into the InstrumentationAdapters. Of course this forces Instrumentation to have dependency on Sampling - but doing this to make progress and uncover any other potential issues.

If everything else is good, and the only remaining issue is Instrumentation being aware of Samplers - we'll try to get a new API in activity source to run the `GetRequestedDataUsingContext`.

Comment (tarekgh) [reply src/OpenTelemetry/Trace/Export/SimpleActivityProcessor.cs]: > Given that making sampling decision in SDK piece is going to be incorrect (as we have incorrect Kind)

Is this to set the Kind? reflection can work here till we expose the Kind setter.

Comment (cijothomas) [reply src/OpenTelemetry/Trace/Export/SimpleActivityProcessor.cs]: We can't set the Kind to the correct one in the OT SDK as SDK doesn't know what is the actual kind. InstrumentationAdapter knows it, and it sets it.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs [hunk 6]
```diff
@@ -66,41 +66,43 @@ public override void OnStartActivity(Activity activity, object payload)
...
+                activity.Context,
+                activity.TraceId,
+                activity.DisplayName,
+                activity.Kind,
+                activity.Tags,
+                activity.Links);
+
+            // TODO: Find a way to avoid Instrumentation being tied to Sampler
+            var samplingDecision = this.sampler.ShouldSample(samplingParameters);
+            activity.IsAllDataRequested = samplingDecision.IsSampled;
```
[the remark this answers is not in the document]
Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:104]: Good point - yes I think. Essentially this should mimic the behavior when `GetRequestedDataUsingContext` was used.

File: src/OpenTelemetry.Instrumentation.AspNetCore/OpenTelemetryBuilderExtensions.cs [hunk 7]
```diff
@@ -39,6 +39,8 @@ public static class OpenTelemetryBuilderExtensions
                 throw new ArgumentNullException(nameof(builder));
             }
 
+            // Asp.Net Core is not instrumented with ActivitySource, hence
+            // it'll have a default ActivitySource with name string.Empty.
             builder.AddActivitySource(string.Empty);
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Instrumentation.AspNetCore/OpenTelemetryBuilderExtensions.cs:44]: @cijothomas Question... if we listen for string.Empty in theory we'll get activities from a bunch of legacy DiagnosticSources flowing through our listener, do we need to filter out the stuff we don't care about somewhere?

File: src/OpenTelemetry.Instrumentation.AspNetCore/OpenTelemetryBuilderExtensions.cs [hunk 8]
```diff
@@ -39,6 +39,8 @@ public static class OpenTelemetryBuilderExtensions
                 throw new ArgumentNullException(nameof(builder));
             }
 
+            // Asp.Net Core is not instrumented with ActivitySource, hence
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Instrumentation.AspNetCore/OpenTelemetryBuilderExtensions.cs:42]: @cijothomas Question... if we listen for string.Empty in theory we'll get activities from a bunch of legacy DiagnosticSources flowing through our listener, do we need to filter out the stuff we don't care about somewhere?

Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.AspNetCore/OpenTelemetryBuilderExtensions.cs:42]: You are right. This will be followed up.

File: samples/Exporters/Web/Startup.cs [hunk 9]
```diff
@@ -36,18 +37,15 @@ public void ConfigureServices(IServiceCollection services)
...
-                    {
-                        options.ServiceName = "test-zipkin";
-                        options.Endpoint = new Uri(this.Configuration.GetValue<string>("Zipkin:Endpoint"));
-                    })
-                    .AddRequestInstrumentation()
-                    .AddDependencyInstrumentation();
-            });
+            OpenTelemetrySdk.Default.EnableOpenTelemetry(
+                (builder) => builder.AddRequestInstrumentation().AddDependencyInstrumentation()
+                .UseJaegerActivityExporter(o =>
```
Comment (cijothomas) [inline samples/Exporters/Web/Startup.cs:42]: moved to jaeger as zipkin exporter with activity is not yet ready

File: samples/Exporters/Web/Controllers/WeatherForecastController.cs [hunk 10]
```diff
@@ -14,9 +16,12 @@ public class WeatherForecastController : ControllerBase
             "Freezing", "Bracing", "Chilly", "Cool", "Mild", "Warm", "Balmy", "Hot", "Sweltering", "Scorching"
         };
 
+        private static HttpClient httpClient = new HttpClient();
+
         [HttpGet]
         public IEnumerable<WeatherForecast> Get()
         {
+            var res = httpClient.GetStringAsync("http://google.com").Result;
```
Comment (pjanotti) [inline samples/Exporters/Web/Controllers/WeatherForecastController.cs:24]: I assume this is only for tests and not part of the final merge.

File: samples/Exporters/Web/OpenTelemetry.Exporter.Web.csproj [hunk 11]
```diff
@@ -12,9 +12,10 @@
   </ItemGroup>
 
   <ItemGroup>
+    <ProjectReference Include="..\..\..\src\OpenTelemetry.Exporter.Console\OpenTelemetry.Exporter.Console.csproj" />
+    <ProjectReference Include="..\..\..\src\OpenTelemetry.Exporter.Jaeger\OpenTelemetry.Exporter.Jaeger.csproj" />
     <ProjectReference Include="..\..\..\src\OpenTelemetry.Instrumentation.AspNetCore\OpenTelemetry.Instrumentation.AspNetCore.csproj" />
     <ProjectReference Include="..\..\..\src\OpenTelemetry.Instrumentation.Dependencies\OpenTelemetry.Instrumentation.Dependencies.csproj" />
-    <ProjectReference Include="..\..\..\src\OpenTelemetry.Exporter.Zipkin\OpenTelemetry.Exporter.Zipkin.csproj" />
```
Comment (pjanotti) [inline samples/Exporters/Web/OpenTelemetry.Exporter.Web.csproj:17]: I will create the Activity exporter for Zipkin this week or the next.

File: samples/Exporters/Web/appsettings.json [hunk 12]
```diff
@@ -7,7 +7,9 @@
     }
   },
   "AllowedHosts": "*",
-  "Zipkin": {
-    "Endpoint": "http://localhost:9411/api/v2/spans"
+  "Jaeger": {
+    "ServiceName": "jaeger-test",
+    "Host": "localhost",
```
Comment (pjanotti) [inline samples/Exporters/Web/appsettings.json:12]: Side note: it feels strange that we need to set Jaeger Host and Port to what it should be default values. Anyway, to be addressed in a separate PR.

File: src/OpenTelemetry.Instrumentation.AspNet.Win/Implementation/HttpInListener.cs [hunk 13]
```diff
@@ -48,53 +53,75 @@ public override void OnStartActivity(Activity activity, object payload)
...
+
             var request = context.Request;
             var requestValues = request.Unvalidated;
 
             // see the spec https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-semantic-conventions.md
             var path = requestValues.Path;
-
-            TelemetrySpan span;
-            if (this.options.TextFormat is TraceContextFormat)
+            activity.DisplayName = path;
```
Comment (pjanotti) [inline src/OpenTelemetry.Instrumentation.AspNet.Win/Implementation/HttpInListener.cs:70]: Due to the risk of high-cardinality on the path OTel spec suggests using the HTTP method instead. That said the path is much more useful in general and having an option to add it (more likely the absolute path) will be useful. There are other locations also using the path as DisplayName, but other correctly use the helper that gets the name per OTel spec. BTW, the link on the comment above is broken.

File: samples/Exporters/Web/Controllers/WeatherForecastController.cs (hunk 10 above)
Comment (cijothomas) [reply samples/Exporters/Web/Controllers/WeatherForecastController.cs:24]: Yes adding a comment.

File: src/OpenTelemetry.Instrumentation.AspNet.Win/Implementation/HttpInListener.cs (hunk 13 above)
Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.AspNet.Win/Implementation/HttpInListener.cs:70]: Yes incoming requests were still using path as SpanName, hence I used it as activity.DisplayName.
This shouldn't block this PR, as we can (and should) address this separate from this PR.

File: src/OpenTelemetry/Trace/Configuration/OpenTelemetrySdk.cs [hunk 14]
```diff
@@ -15,20 +15,29 @@
...
 using System.Diagnostics;
 using OpenTelemetry.Trace.Export;
 using OpenTelemetry.Trace.Samplers;
 
 namespace OpenTelemetry.Trace.Configuration
 {
-    public class OpenTelemetrySdk
+    public class OpenTelemetrySdk : IDisposable
     {
+        public static OpenTelemetrySdk Default = new OpenTelemetrySdk();
```
[the remark this answers is not in the document]
Comment (pjanotti) [reply src/OpenTelemetry/Trace/Configuration/OpenTelemetrySdk.cs:27]: I'm not 100% sure but I tend to like your approach @CodeBlanch but I think we can address it in a separate PR/discussion.

[the remark this answers is not in the document]
Comment (cijothomas) [reply src/OpenTelemetry/Trace/Configuration/OpenTelemetrySdk.cs:27]: @CodeBlanch shared another proposal which sounds very good! I will add it as a follow up PR.
